Handler map localization

parthux1 opened the pull request:
# PR for Issue #43.

## Structure
1. added folder `locales` containing
1.1  [common.py](https://github.com/parthux1/google_takeout_parser/blob/bddaaeb5346a5f9f132e9358e99d13eff9629835/google_takeout_parser/locales/common.py) - defines several TakeoutFiles(Enum), which are resolved to a parser later
1.2 [en.py](https://github.com/parthux1/google_takeout_parser/blob/bddaaeb5346a5f9f132e9358e99d13eff9629835/google_takeout_parser/locales/en.py), [de.py](https://github.com/parthux1/google_takeout_parser/blob/handlerMap_localization/google_takeout_parser/locales/de.py) - HandlerMap from path_dispatch moved here. Each module defines such a map and links filepaths to TakeoutFiles from common.py

2. added [path_handler.py](https://github.com/parthux1/google_takeout_parser/blob/bddaaeb5346a5f9f132e9358e99d13eff9629835/google_takeout_parser/path_handler.py) which
2.1 resolves enum `TakeoutFile` to a `callable` (prev. happened directly in the HandlerMap)
2.2 groups localized HandlerMaps (`en.py, de.py`) into the class `Localizedhandlers`

> due to changes on class `TakeoutParser` you can e.g do 
```python
from google_takeout_parser.path_dispatch import TakeoutParser, LocalizedHandler
tp = TakeoutParser('path/to/takeout', handlers=[LocalizedHandler.EN(), LocalizedHandler.DE()])
```
> this would try to map each encountered file to the EN map, and if not resolved to the DE map afterwards.

Details following:

## Changes to existing code
- [TakeoutParser now defaults to EN handler](https://github.com/parthux1/google_takeout_parser/blob/bddaaeb5346a5f9f132e9358e99d13eff9629835/google_takeout_parser/path_dispatch.py#L100). an undefined amount of HandlerMaps can be supplied
- [removed default-handler from dispatch map](https://github.com/parthux1/google_takeout_parser/blob/bddaaeb5346a5f9f132e9358e99d13eff9629835/google_takeout_parser/path_dispatch.py#L151). The function now iterates over all supplied handlers

- previously a HandlerMap mapped to either `None` or `callable`. Now if you map to `TakeoutFile` [path_handler.py will parse it to a `callable`](https://github.com/parthux1/google_takeout_parser/blob/bddaaeb5346a5f9f132e9358e99d13eff9629835/google_takeout_parser/path_dispatch.py#L128)

> I think these features compliment each other pretty well. One could define some kind of "Standard-Ignore" HandlerMap which is supplied as the last element to `TakeoutParser`.
> Custom functions could be given different priorities by list-order

## current problems:
1. parsing my-activity files yield a warning eg. "time data '12.07.2019, 14:51:43' does not match format '%b %d, %Y, %I:%M:%S %p"
2. `__main__.py` was updated but doesn't handle the parser flag because I have no idea how to debug it. Currently I debug by importing classes as shown in the prev. example
3. [TakeoutParser._warn_if_no_activity](https://github.com/seanbreckenridge/google_takeout_parser/blob/368017199f822196f78a3e70f7b27584066b8796/google_takeout_parser/path_dispatch.py#L212) path needs to be resolved via the localized HandlerMap. One could just hardcode a key to access the map, but I don't really like this solution
  
---------------
Is this concept understandable and beneficial?
Are there any files I should keep in mind while working?
How should I start adding tests for this?

seanbreckenridge commented on google_takeout_parser/locales/common.py:
```diff
@@ -0,0 +1,27 @@
+    Optional
+)
+
+class TakeoutFile(Enum):
+    CHROME_HISTORY = auto(),
+    GPLAYSTORE_INSTALLS = auto(),
+    LOCATION_HISTORY = auto(),
+    LOCATION_HISTORY_SEMANTIC = auto(),
+    YOUTUBE_HISTORY_HTML = auto(),
+    YOUTUBE_HISTORY_JSON = auto(),
+    YOUTUBE_COMMENT = auto(),
+    YOUTUBE_COMMENT_LIVECHAT = auto(),
+    YOUTUBE_LIKES = auto(),
+    # Parent Folder "My Activity"
+    ACTIVITY_HTML = auto(),
+    ACTIVITY_JSON = auto()
+
+HandlerMap = Dict[str, Optional[TakeoutFile]]
```
Hmm, I sort of see this enum adding more problems than it solves. It means the user has to redefine/subclass this enum if they ever want to add a custom file this parses without just forking and editing this directly, and they also have to edit the global variable `TAKEOUT_PARSER` to add their function there

It does mean you have to import the functions at the top of every locale file, but I think I would actually prefer that.

Smaller issue is that this adds indirection in that you have to know about `.common.py` and the `TAKEOUT_PARSER` map. If someone was exploring this and the functions were inlined here, its much easier to jump from `locales/de.py` to the `_parse_json_activity` function than having to explore to make sure you're going to the correct `Enum` value from the `TAKEOUT_PARSER` dict

There is a bit of personal bias here in that `Enum`s are not as powerful in python than in other languages so unless I have a good reason to use them I won't, but unless you have some reason not to I think it'd be better to just put the functions directly in the handler map like they were before

parthux1 replied:
After some time thinking about it I generally agree with mapping to functions directly.  

All in the layer of indirection only allows changing parser-functions for all locales at once using the `ParserMap`.
When adding a new locale one would also look at `en.py` anyway. Doesn't matter if you copy functions or Enums from there.

I still want to present my thought-process so we're on the same page.  

If you still agree on including functions directly into localized maps I'll do it, leaving only type definitions and the `LocalizedHandler` handler class in `locale/common.py`.

also there's actually a circular import in the `locale`-Files:
- common.py imports all langs to provide them in `LocalizedHandler`
- each lang imports common.py to get the typedef `HandlerMap`

***Is this a problem when working with `from x import y`?***

### my thought process

In my opinion a localized handler map should only account for the localization-differences, in this case meaning   
  
1. A HandlerMap maps the locale name to a **general file name**, currently implemented as the Enum `TakeoutFile`
2. The ParserMap [TAKEOUT_PARSER](https://github.com/parthux1/google_takeout_parser/blob/5d2c905afba3718e5858632398d95692bd2f031b/google_takeout_parser/path_handler.py#L35) (should definetly rename it. TakeoutParser is already a class) knows how to parse **general file names**

Using this concept each locale would have to map their custom regexes to a "standardized" file type.

### Mentioned problem: adding custom parsers
> It means the user has to redefine/subclass this enum if they ever want to add a custom file this parses without just forking and editing this directly, and they also have to edit the global variable TAKEOUT_PARSER to add their function there

The current implementation will try to parse a file by iterating over all parsers.
[dispatch_map](https://github.com/parthux1/google_takeout_parser/blob/5d2c905afba3718e5858632398d95692bd2f031b/google_takeout_parser/path_dispatch.py#L140) allow passing functions.
By supplying a custom map you could add custom parsers:
1. Put all own parsers, as well as their "trigger-regex" in a custom `HandlerMap`
2. Define a list containing the custom `handlerMap` before any standard-HandlerMap
4. Supply this list to `TakeoutParser`
```python
my_map: HandlerMap = {r"regex/to/file": my_handler}
t = TakeoutParser("path/to/takeout", handlers=[my_map, LocalizedHandler.EN()])
```

seanbreckenridge replied:
Regarding the circular import, you could instead move the `LocalizedHandler` which collects all the maps to something like an `locales/all.py` file, and only leave the types/shared functions in `common.py`. 

> custom parsers

Oh, yeah, if the enum is removed there is no issue with providing a custom parser, your though process makes sense there.

seanbreckenridge commented on google_takeout_parser/__main__.py:
```diff
@@ -40,6 +40,15 @@ def main(verbose: Optional[bool]) -> None:
+    click.option(
+        "-l",
+        "--localization",
+        type=click.Choice(["en", "ger"]),
```
You probably know better than I do, is it common for this to be `ger` or `de`?

seanbreckenridge commented on google_takeout_parser/path_dispatch.py:
```diff
@@ -230,7 +135,15 @@ def _match_handler(p: Path, handler: HandlerMap) -> HandlerMatch:
+                    return None
+                elif type(h) is TakeoutFile:
+                    return TAKEOUT_PARSER[h] # resolve TakeoutFile to a parser function
+                elif type(h) is callable:
```
Should do `elif callable(h)` instead

seanbreckenridge commented:
>  Are there any files I should keep in mind while working?

For `__main__.py`, incase it wasnt obvious since the decorator is somewhat complicated, to add the CLI flag, it'd be:

```diff
diff --git a/google_takeout_parser/__main__.py b/google_takeout_parser/__main__.py
index d5fe165..31cd5e8 100644
--- a/google_takeout_parser/__main__.py
+++ b/google_takeout_parser/__main__.py
@@ -43,7 +43,7 @@ SHARED = [
     # TODO: actually handle argument
     click.option(
         "-l",
-        "--localization",
+        "--locale",
         type=click.Choice(["en", "ger"]),
         default="en",
         help="Used DEFAULT_HANDLER_MAP resoling folder names to parser models",
@@ -110,7 +110,7 @@ def parse(cache: bool, action: str, takeout_dir: str) -> None:
 @main.command(short_help="merge multiple takeout directories")
 @shared_options
 @click.argument("TAKEOUT_DIR", type=click.Path(exists=True), nargs=-1, required=True)
-def merge(cache: bool, action: str, takeout_dir: Sequence[str]) -> None:
+def merge(cache: bool, action: str, locale: str, takeout_dir: Sequence[str]) -> None:
     """
     Parse and merge multiple takeout directories
     """
```

Also I think should rename to `--locale`, shorter to type

>  How should I start adding tests for this?

For the `path_handler` you've done what I would've synced a copy of the directory -- good.

Otherwise, I think anywhere where you see differences in the actual structure of code where it leads to differences, you could extract an HTML div or JSON blob like I do in:

- https://github.com/seanbreckenridge/google_takeout_parser/blob/master/tests/test_json.py
- https://github.com/seanbreckenridge/google_takeout_parser/blob/master/google_takeout_parser/parse_html/test_html_parsing.py

I dont think there should be many of those though, but if you're seeing errors there might be a differently structured file for different locales

> parsing my-activity files yield a warning eg. "time data '12.07.2019, 14:51:43' does not match format '%b %d, %Y, %I:%M:%S %p"

I think that error in particular is the localizations of the dates when using the html exports (e.g. for the `My Activity.html`)

Those are *pretty annoying* bits of code since google takeout has changed a bunch and we have to account for all the changes (at least in the english locale).

I think those should be from `My Acitvity.html`, so unless you want to pass the locale to `parse_html_dt`, we could opt to just not include the HTML parsing code for other locales. The code to parse the HTML should be seen as legacy anyways -- as it says in the [readme](https://github.com/seanbreckenridge/google_takeout_parser) you should pick `JSON` for those formats, as it then just returns an UTC epoch timestamp instead of having to handle dealing with locales. Could warn the user to do exactly that if parsing html files with other locales

You should also just make sure none of the other html parsing code is incorrect in a different locale. E.g. does [comment.py](https://github.com/seanbreckenridge/google_takeout_parser/blob/master/google_takeout_parser/parse_html/comment.py) have differences in german or is the date format the same?

seanbreckenridge commented on google_takeout_parser/path_handler.py:
```diff
@@ -0,0 +1,63 @@
+
+class LocalizedHandler:
+    
+    # static methods resolving to a localized HandlerMap
+    def EN() -> HandlerMap:
+        return hmap_en
```
should annotate these with `@staticmethod` https://docs.python.org/3/library/functions.html#staticmethod

parthux1 commented:
Several change requests marked as resolved are implemented as proposed by you.

## I left 2 discussions open:
1. [Resolve File to Parser in dispatch_map](https://github.com/seanbreckenridge/google_takeout_parser/pull/44#discussion_r1155173732) I changed the var name and switched the actually not working as intended `continue` to `break`. Should be okay now but you may want to take a quick look.
2. [Switch Enum-Mapping to referencing the parser functions in each locale HandlerMap](https://github.com/seanbreckenridge/google_takeout_parser/pull/44#discussion_r1155175756) Let's discuss this in the corresponding thread.

## what I still need to do:
1. Think about the date-localization problem. I'm currently getting another takeout with jsons to check this.  
Will probably just supply some readme section about this and reference it in the warning.
2. Check for other differences like 1. and add tests to the corresponding `tests/test_`-files
2. Before merging: update readmes "Library Usage" section

## Regarding `__main__.py`
How do you test the command line behavior?😅  
When trying to call it via command line 
```bash
python google_takeout_parser\__main__.py parse [File]
```
I get errors due to the relative `.` imports. Do I have to locally package it in some way?

seanbreckenridge commented:
> How do you test the command line behavior

While in the the root directory, run:

```
pip install .
```

to install it, and then you can use `python3 -m google_takeout_parser`. You could also just run `python3 -m google_takeout_parser` without the `pip install .`, but I tend to install it just to make sure Im not mistakenly in the wrong directory and using the globally installed `google_takeout_parser` and not the local version

seanbreckenridge commented on google_takeout_parser/path_dispatch.py:
```diff
@@ -244,35 +157,32 @@ def dispatch_map(self) -> Dict[Path, HandlerFunction]:
+                    continue
 
-            # don't raise errors here since the DEFAULT_HANDLER_MAP may handle parsing it
+                last_handler = file_handler
```
Will check this out and test some time this week, bit too hard for me to reason through in the github PR
